data/web/sogo-auth.php

```diff
@@ -89,6 +89,13 @@
     header("X-Auth: Basic ".base64_encode("$username:$password"));
     header("X-Auth-Type: Basic");
     exit;
+  } elseif(isset($_SESSION['mailcow_cc_username'])) {
```

Comment:
As far as I know, mailcow_cc_username will not be a valid email if the user is admin / domain admin.
So the current proposal would essentially send invalid basic_auth credentials to SOGo in some cases / not work for the ADMIN_EMAIL_LOGIN feature.

Maybe the following would be more concise:
```
} elseif (
      is_array($_SESSION[$session_var_user_allowed]) &&
      count($_SESSION[$session_var_user_allowed]) === 1
) {
    $username = $_SESSION[$session_var_user_allowed][0];
```

This would prevent sending invalid credentials to SOGo, and as long as admins are only logged into one mailbox, should work as expected.
The workaround will still stop working as soon as an admin opens multiple mailboxes, but would at least skip this check / send empty authentication headers as before.

Reply:
I thought about the admin login too, but since even your suggestion only fixes the problem as long as only a single mailbox is opened, I don‘t think it’s worth bothering with. I’ll add a note to https://github.com/mailcow/mailcow-dockerized-docs/blob/Master/docs/debug-admin_login_sogo.md. Instead. Also note that I found that `$_ SESSION[$session_var_user_allowed]` sometimes contains the same email address multiple times, so it would need to be deduplicated before doing as you suggest.

Reply:
Also need remove or rephrase test in `debug-admin_login_sogo` doc:
`All of these options / behaviors are disabled if the ALLOW_ADMIN_EMAIL_LOGIN is not enabled in the config.`
`Drawbacks when enabled`
because now any user have `login to web` button which works in mostly same way.

Reply:
In general the change looks ok (besides the concern below). What `sogo-auth.php` does is telling Sogo what mailbox to use when calling its REST endpoints. In addition we also provide the mailbox password that Sogo uses to login to dovecot (both items are provided as X-headers).

I do have one concern with the change: The code in GET login checks if `$login` has access and adds it to the list only then. Using `$_SESSION['mailcow_cc_username']` directly without checks might be dangerous. How about:
```php
  // extract email address from "/SOGo/so/user@domain/xy"
  $url_parts = explode("/", $_SERVER['HTTP_X_ORIGINAL_URI']);
  $email_list = [
    $url_parts[3],                           // Requested mailbox
    ($_SESSION['mailcow_cc_username'] ?? '') // Current user
  ];

  foreach ($email_list as $email) {
    // check if this email is in session allowed list
    if (
        !empty($email) &&
        filter_var($email, FILTER_VALIDATE_EMAIL) &&
        is_array($_SESSION[$session_var_user_allowed]) &&
        in_array($email, $_SESSION[$session_var_user_allowed])
    ) {
      $username = $email;
      $password = $_SESSION[$session_var_pass];
      header("X-User: $username");
      header("X-Auth: Basic ".base64_encode("$username:$password"));
      header("X-Auth-Type: Basic");
      exit;
    }
  }
```
This still prefers the requested mailbox and falls back to the current user (when it is a "user" account where login is email address). All existing checks remain in place since the condition only succeeds if the current user session has access.

Reply:
Thanks for taking a look, @jkellerer. `$_SESSION['mailcow_cc_username']` is always present in `$_SESSION[$session_var_user_allowed]`, so I don't understand how your code would change anything. I don't think `$_SESSION['mailcow_cc_username']` might be dangerous -- it's a session variable, so we can trust that it always contains the name of the logged-in user and cannot be overwritten by the client.

Reply:
Hi @mkuron, yes I agree to your statement, however your change adds a separate logic branch by always accepting `$_SESSION['mailcow_cc_username']` for SSO without requiring to go through `if (isset($_GET['login'])) { ... }` (the explicit request to SSO to SOGo). The only reason why SSO to SOGo still requires the `$_GET['login']` logic for an ordinary user is that `$_SESSION[$session_var_pass]` is unset at first but this is a side effect. 

For that reason I think it would be cleaner to keep it separated: 
* The first logic branch (`$_GET['login']` logic) fills the array of allowed mail accounts for SSO for the current user
* The second logic branch handles auth requests for SOGo on a list of possible mail-accounts for a request. Either the ones explicitly requested in the URL, or the user's own (possibly including the latest dual login). 
* Only the first branch needs to deal with "is it allowed" the second only needs to interpret which mailbox is requested (concerns are separated).

E.g. by extending `$email_list` you can even support the latest dual login of the admin:
```php
  $email_list = [
    $url_parts[3],                               // Requested mailbox
    ($_SESSION['mailcow_cc_username'] ?? '')     // Current user
    ($_SESSION["dual-login"]["username"] ?? '')  // Dual login user
  ];
```

Reply:
@jkellerer, thanks for the explanation. Your solution is indeed better and I have updated my pull request.